### DISCUSSION TITLE
Rewrite request Host header

### DISCUSSION THREAD

**Discussion by jRom42** [membrane/api-gateway#1404] [General]
Hello,

A backend web service that used to work for years has stopped working because it looks like a control is now made on the Host header in the request.
The web service still works with a curl command when accessed without membrane.

This call works
```bash
curl --request POST   --url https://<**external_webservice_fqdn**>/service   \
  --header 'accept: application/json'   \
  --header 'content-type: application/x-www-form-urlencoded'     \  
  --data username=xxxxxxxxx   \
  --data password=xxxxxxxxxx
```
 
This call does not work anymore:

```bash
curl --request POST   --url https://<**membrane_fqdn**>/service   \
  --header 'accept: application/json'   \
  --header 'content-type: application/x-www-form-urlencoded'     \  
  --data username=xxxxxxxxx   \
  --data password=xxxxxxxxxx
```

My membrane version is 5.5.9

Service configuration is very simple:

```
<serviceProxy name="myservice" host="membrane.xxx.xx">
        <path>/service/path</path>


        <target url="https://<backend_server>">
                <ssl>
                        <truststore location="<jks_path>" password="xxxxxxxx"/>
                </ssl>
        </target>
</serviceProxy>
```

How can I overwrite the Host header before the request is sent to the target ?

I have tried with **_groovy_** and **_setHeader_** but none of them seem to work.

```
<request>
    <setHeader name="Host" value="xxxxxx" />
</request>
```

```
<groovy>
    exc.getRequest().getHeader().setHost("xxxxx");
    CONTINUE
</groovy>
```

Web service response

![image](https://github.com/user-attachments/assets/62fc01e3-32c5-42b2-896b-c14234f12a80)

![image](https://github.com/user-attachments/assets/c2df9830-9886-4e94-912b-9c3967360e39)

**Comment by predic8** [comment]
Hi,
Membrane automatically sets the host header according to the target element. You can turn that off by setting an attribute on router: 

<router adjustHostHeader="false">

See: https://www.membrane-api.io/docs/current/router.html

Then your code snippets should work. But usually Membrane sets it to the right value. I would first check how the headers are set using a log inside your serviceProxy:

```
<serviceProxy name="myservice" host="membrane.xxx.xx">
        <path>/service/path</path>
        <log/>
        <target url="https://<backend_server>">
                <ssl>
                        <truststore location="<jks_path>" password="xxxxxxxx"/>
                </ssl>
        </target>
</serviceProxy>
```

Then make the call and have a look at the console.

Cheers
Thomas

**Comment by jRom42** [comment]
Hi Thomas,

Thank you for taking time to answer.
It works with adjustHostHeader="false" on serviceProxy (setting on router membrane fails with an error) and this groovy config:

```
<groovy>
    exc.getRequest().getHeader().setHost("xxxxx");
    CONTINUE
</groovy>
```
